app/ota_client.py

```diff
@@ -873,177 +1062,136 @@ def _create_symbolic_links(self, list_file, standby_path):
-            """
-            processed_list have dictionaries as follows:
-            {"size": int}  # file size
-            {"elapsed": int}  # elapsed time in seconds
-            {"op": str}  # operation. "copy", "link" or "download"
-            {"errors": int}  # number of errors that occurred when downloading.
-            """
-            processed_list = manager.list()
-
-            def error_callback(e):
-                error_queue.put(e)
-
-            boot_standby_path = self.get_standby_boot_partition_path()
-            # bind the required options before we use this method
-            _create_regfile_func = partial(
-                self._create_regular_file,
-                url_base=url_base,
-                cookies=cookies,
-                standby_path=standby_path,
-                processed_list=processed_list,
-                boot_standby_path=boot_standby_path,
-                downloader=self._download,
-            )
-
-            # NOTE: temporary limit the max worker to not greater than 3
-            with Pool(processes=min(3, os.cpu_count())) as pool:
-                hardlink_dict = dict()  # sha256hash[tuple[reginf, event]
-
-                with open(list_file, "r") as f:
-                    for reginf in map(RegularInf, f):
-                        if reginf.nlink >= 2:
-                            prev_reginf, event = hardlink_dict.setdefault(
-                                reginf.sha256hash, (reginf, manager.Event())
-                            )
-
-                            # multiprocessing.apply_async
-                            # input args:
-                            #   func, args: list, kwargs: dict, *, callback, error_callback
-                            # output:
-                            #   async_result
-                            pool.apply_async(
-                                _create_regfile_func,
-                                (reginf, prev_reginf),
-                                {"hardlink_event": event},
-                                error_callback=error_callback,
-                            )
-                        else:
-                            pool.apply_async(
-                                _create_regfile_func,
-                                (reginf,),
-                                error_callback=error_callback,
-                            )
-
-                pool.close()
-                while len(processed_list) < regular_files_num:
-                    self._set_statistics(processed_list)
+        self._statistics.set("total_regular_files", total_files_num)
+
+        _hardlink_register = _HardlinkRegister()
+        # maximum allowd concurrent tasks
+        _se = threading.Semaphore(self.MAX_CONCURRENT_DOWNLOAD_TASKS)
+        # collector that records stats from tasks and update ota-update status
+        _collector = _CreateRegularStatsCollector(
+            self._statistics, _se, interval=self.COLLECT_INTERVAL
+        )
```

Comment:
I'd like to remove relationship between _CreateRegularStatsCollector and _se.
Like:
```
def done_callback(se):
  _collector.callback()
  se.release()
...
fut.add_done_callback(done_callback)
```
Because _se.acquire() is done in the loop, but _se.release() is done in the _CreateRegularStatsCollector class.
To understand control flow easier.

Reply:
Sure! I will let the callback to take `_se` as input param and release it.